File: test/python/node_actions_test.py
from unittest import TestCase
from .parse_helper import ParseHelper
from grammars import node_actions


class NodeActionsTest(TestCase, ParseHelper):
    def test_makes_nodes_from_a_string(self):
        input  = "act-str: hello"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["str", input, 9, 14, []], result)

    def test_makes_nodes_from_a_char_class(self):
        input  = "act-class: k"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["char", input, 11, 12, []], result)

    def test_makes_nodes_from_any_char(self):
        input  = "act-any: ?"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["any", input, 9, 10, []], result)

    def test_makes_nodes_from_a_maybe_rule(self):
        input  = "act-maybe: hello"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["maybe", input, 11, 16, []], result)

    def test_does_not_invoke_an_action_for_a_maybe_rule_with_no_match(self):
        input  = "act-maybe: "
        self.assertParse(("", 11, []), node_actions.parse(input, actions=TestActions()))

    def test_makes_nodes_from_a_repetition(self):
        input  = "act-rep: abc"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["rep", input, 9, 12], result[0:4])

        self.assertParseElements([
            ("a", 9, []),
            ("b", 10, []),
            ("c", 11, [])
        ], result)

    def test_makes_nodes_from_a_sequence(self):
        input  = "act-seq: xyz"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["seq", input, 9, 12], result[0:4])

        self.assertParseElements([
            ("x", 9, []),
            ("y", 10, []),
            ("z", 11, [])
        ], result)

    def test_makes_nodes_from_a_parenthesised_expression(self):
        input  = "act-paren: !"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["paren", input, 11, 12, []], result)

    def test_binds_to_the_options_of_a_choice(self):
        input  = "act-choice: 0"
        result = node_actions.parse(input, actions=TestActions()).elements[1]
        self.assertEqual(["zero", input, 12, 13, []], result)

        input  = "act-choice: 42"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(["int", input, 12, 14], result[0:4])

        self.assertParseElements([
            ("4", 12, []),
            ("2", 13, [
                ("2", 13, [])
            ])
        ], result)

    def test_treats_null_as_a_valid_result(self):
        result = node_actions.parse("act-falsey: null", actions=TestActions()).elements[1]
        self.assertEqual(None, result)

    def test_treats_false_as_a_valid_result(self):
        result = node_actions.parse("act-falsey: false", actions=TestActions()).elements[1]
        self.assertEqual(False, result)

    def test_treats_zero_as_a_valid_result(self):
        result = node_actions.parse("act-falsey: 0", actions=TestActions()).elements[1]
        self.assertEqual(0, result)

    def test_treats_empty_strings_as_a_valid_result(self):
        result = node_actions.parse("act-falsey: ''", actions=TestActions()).elements[1]
        self.assertEqual("", result)

    def test_treats_empty_lists_as_a_valid_result(self):
        result = node_actions.parse("act-falsey: []", actions=TestActions()).elements[1]
        self.assertEqual([], result)

    def test_treats_falsey_values_as_acceptable_lookahead_results(self):
        input  = "act-falsey-pred: 0"
        result = node_actions.parse(input, actions=TestActions()).elements[1].elements[1]

        self.assertEqual(["zero", input, 17, 18, []], result)

    def test_treats_falsey_values_as_acceptable_repetition_results(self):
        result = node_actions.parse("act-falsey-rep: null0false''[]", actions=TestActions())
        self.assertEqual([None, 0, False, "", []], result.elements[1].elements)

    def test_treats_falsey_values_as_acceptable_maybe_results(self):
        input  = "act-falsey-opt: null"
        result = node_actions.parse(input, actions=TestActions()).elements[1]

        self.assertEqual(None, result)


class TestActions:
    def make_str(self, *args):
        return ["str"] + list(args)

    def make_char(self, *args):
        return ["char"] + list(args)

    def make_any(self, *args):
        return ["any"] + list(args)

    def make_maybe(self, *args):
        return ["maybe"] + list(args)

    def make_rep(self, *args):
        return ["rep"] + list(args)

    def make_seq(self, *args):
        return ["seq"] + list(args)

    def make_paren(self, *args):
        return ["paren"] + list(args)

    def make_zero(self, *args):
        return ["zero"] + list(args)

    def make_int(self, *args):
        return ["int"] + list(args)

    def make_null(*args):
        return None

    def make_false(*args):
        return False

    def make_0(*args):
        return 0

    def make_empty_str(*args):
        return ""

    def make_empty_list(*args):
        return []

File: test/cs/NodeActionsTest.cs
//package canopy.node_actions;

using System.Collections.Generic;
using System.Collections;

using Microsoft.VisualStudio.TestTools.UnitTesting;

//import helpers.ElementsSpec;
//import helpers.Node;
//import helpers.NodeSpec;

using test.grammars.node_actions.Actions;
using test.grammars.node_actions.Label;
using test.grammars.node_actions.ParseError;
using test.grammars.node_actions.NodeActions;
using test.grammars.node_actions.TreeNode;
[TestClass]
class NodeActionsTest : ParseHelper {
    [TestMethod]
    void makesNodesFromAString(){
        String input = "act-str: hello";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(new CustomNode("str", input, 9, 14).noElems(), result);
    }

    [TestMethod]
    void makesNodesFromACharClass(){
        String input = "act-class: k";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(new CustomNode("char", input, 11, 12).noElems(), result);
    }

    [TestMethod]
    void makesNodesFromAnyChar(){
        String input = "act-any: ?";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(new CustomNode("any", input, 9, 10).noElems(), result);
    }

    [TestMethod]
    void makesNodesFromAMaybeRule(){
        String input = "act-maybe: hello";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(new CustomNode("maybe", input, 11, 16).noElems(), result);
    }

    [TestMethod]
    void doesNotInvokeAnActionForAMaybeRuleWithNoMatch(){
        String input = "act-maybe: ";
        TreeNode result = NodeActions.parse(input, new TestActions());

        expect(result).toMatch(node("", 11).noElems());
    }

    [TestMethod]
    void makesNodesFromARepetition(){
        String input = "act-rep: abc";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(
            new CustomNode("rep", input, 9, 12)
                .elem(node("a", 9).noElems())
                .elem(node("b", 10).noElems())
                .elem(node("c", 11).noElems()),
            result
        );
    }

    [TestMethod]
    void makesNodesFromARepetitionInParentheses(){
        String input = "act-rep-paren: abab";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(
            new CustomNode("rep-paren", input, 15, 19)
                .elem(node("ab", 15)
                    .elem(node("a", 15).noElems())
                    .elem(node("b", 16).noElems())
                )
                .elem(node("ab", 17)
                    .elem(node("a", 17).noElems())
                    .elem(node("b", 18).noElems())
                ),
            result
        );
    }

    [TestMethod]
    void makesNodesFromASequence(){
        String input = "act-seq: xyz";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(
            new CustomNode("seq", input, 9, 12)
                .elem(node("x", 9).noElems())
                .elem(node("y", 10).noElems())
                .elem(node("z", 11).noElems()),
            result
        );
    }

    [TestMethod]
    void makesNodesFromASequenceWithMutedElements(){
        String input = "act-seq-mute: xyz";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(
            new CustomNode("seq", input, 14, 17)
                .elem(node("x", 14).noElems())
                .elem(node("z", 16).noElems()),
            result
        );
    }

    [TestMethod]
    void makesNodesFromAParenthesisedExpression(){
        String input = "act-paren: !";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(new CustomNode("paren", input, 11, 12).noElems(), result);
    }

    [TestMethod]
    void bindsToTheOptionsOfAChoice(){
        String input = "act-choice: 0";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);
        assertNode(new CustomNode("zero", input, 12, 13).noElems(), result);

        input = "act-choice: 42";
        result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1);

        assertNode(
            new CustomNode("int", input, 12, 14)
                .elem(node("4", 12).noElems())
                .elem(node("2", 13)
                    .elem(node("2", 13).noElems())
                ),
            result
        );
    }

    [TestMethod]
    void treatsNullAsAValidResult(){
        CustomNode result = (CustomNode)NodeActions.parse("act-falsey: null", new TestActions()).elements.get(1);
        assertEquals(null, result);
    }

    [TestMethod]
    void treatsFalseAsAValidResult(){
        CustomNode result = (CustomNode)NodeActions.parse("act-falsey: false", new TestActions()).elements.get(1);
        assertEquals(null, result);
    }

    [TestMethod]
    void treatsZeroAsAValidResult(){
        CustomNode result = (CustomNode)NodeActions.parse("act-falsey: 0", new TestActions()).elements.get(1);
        assertEquals(null, result);
    }

    [TestMethod]
    void treatsEmptyStringsAsAValidResult(){
        CustomNode result = (CustomNode)NodeActions.parse("act-falsey: ''", new TestActions()).elements.get(1);
        assertEquals(null, result);
    }

    [TestMethod]
    void treatsEmptyListsAsAValidResult(){
        CustomNode result = (CustomNode)NodeActions.parse("act-falsey: []", new TestActions()).elements.get(1);
        assertEquals(null, result);
    }

    [TestMethod]
    void treatsFalseyValuesAsAcceptableLookaheadResults(){
        String input = "act-falsey-pred: 0";
        CustomNode result = (CustomNode)NodeActions.parse(input, new TestActions()).elements.get(1).elements.get(1);

        assertNode(new CustomNode("zero", input, 17, 18).noElems(), result);
    }

    [TestMethod]
    void treatsFalseyValuesAsAcceptableRepetitionResults(){
        String input = "act-falsey-rep: null0false''[]";
        List<TreeNode> elements = NodeActions.parse(input, new TestActions()).elements.get(1).elements;

        assertEquals(5, elements.size());
        foreach (TreeNode elem in elements) {
            assertEquals(null, elem);
        }
    }

    [TestMethod]
    void treatsFalseyValuesAsAcceptableMaybeResults(){
        String input = "act-falsey-opt: null";
        TreeNode result = NodeActions.parse(input, new TestActions()).elements.get(1);

        assertEquals(null, result);
    }

    [TestMethod]
    void treatsFalseyValuesAsAcceptableSequenceResults(){
        String input = "act-falsey-seq: (null)";
        TreeNode result = NodeActions.parse(input, new TestActions()).elements.get(1);

        assertEquals(3, result.elements.size());
        assertEquals(null, result.elements.get(1));
    }

    [TestMethod]
    void treatsFalseyValuesAsAcceptableChoiceResults(){
        String input = "act-falsey-choice: null";
        TreeNode result = NodeActions.parse(input, new TestActions()).elements.get(1);

        assertEquals(null, result);
    }
}

class TestActions : Actions {
    public CustomNode make_str(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("str", input, start, end, elements);
    }

    public CustomNode make_char(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("char", input, start, end, elements);
    }

    public CustomNode make_any(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("any", input, start, end, elements);
    }

    public CustomNode make_maybe(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("maybe", input, start, end, elements);
    }

    public CustomNode make_rep(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("rep", input, start, end, elements);
    }

    public CustomNode make_seq(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("seq", input, start, end, elements);
    }

    public CustomNode make_paren(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("paren", input, start, end, elements);
    }

    public CustomNode make_rep_paren(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("rep-paren", input, start, end, elements);
    }

    public CustomNode make_zero(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("zero", input, start, end, elements);
    }

    public CustomNode make_int(String input, int start, int end, List<TreeNode> elements) {
        return new CustomNode("int", input, start, end, elements);
    }

    public CustomNode make_null(String input, int start, int end, List<TreeNode> elements) {
        return null;
    }

    public CustomNode make_false(String input, int start, int end, List<TreeNode> elements) {
        return null;
    }

    public CustomNode make_0(String input, int start, int end, List<TreeNode> elements) {
        return null;
    }

    public CustomNode make_empty_str(String input, int start, int end, List<TreeNode> elements) {
        return null;
    }

    public CustomNode make_empty_list(String input, int start, int end, List<TreeNode> elements) {
        return null;
    }
}

class CustomNode : TreeNode {
    String type, input;
    int start, end;
    List<TreeNode> elements;
    ElementsSpec<Label> elemsSpec = new ElementsSpec<Label>();

    CustomNode(String type, String input, int start, int end) {
        this(type, input, start, end, null);
    }

    CustomNode(String type, String input, int start, int end, List<TreeNode> elements) {
        this.type = type;
        this.input = input;
        this.start = start;
        this.end = end;
        this.elements = elements;
    }

    CustomNode noElems() {
        elemsSpec.noElems();
        return this;
    }

    CustomNode elem(NodeSpec<Label> elem) {
        elemsSpec.elem(elem);
        return this;
    }
}

class ParseHelper {
    Node<Label> expect(TreeNode node) {
        return new NodeWrapper(node.elements.get(1));
    }

    NodeSpec<Label> node(String text, int offset) {
        return new NodeSpec<Label>(text, offset);
    }

    void assertNode(CustomNode expected, CustomNode actual) {
        assertEquals(expected.type, actual.type);
        assertEquals(expected.input, actual.input);
        assertEquals(expected.start, actual.start);
        assertEquals(expected.end, actual.end);
        expected.elemsSpec.check(new NodeWrapper(new TreeNode("", 0, actual.elements)));
    }
}

class NodeWrapper : Node<Label> {
    private TreeNode node;

    NodeWrapper(TreeNode node) {
        this.node = node;
    }

    public String text() {
        return node.text;
    }

    public int offset() {
        return node.offset;
    }

    public List<Node<Label>> elements() {
        List<Node<Label>> ret = new List<Node<Label>>();
        foreach (var item in node.elements) {
            ret.Add(new NodeWrapper(item));
        }
        return ret;
    }

    public Node<Label> get(Label key) {
        return new NodeWrapper(node.get(key));
    }
}
